x-pack/dev-tools/simulate_autoscaling.py

```diff
@@ -0,0 +1,276 @@
+
+
+START_TIME = 0
+END_TIME = 5*3600
+MEASUREMENT_INTERVAL = 1
+
+AUTO_SCALE_UP_THRESHOLD = 0.9
+AUTO_SCALE_DOWN_THRESHOLD = 0.8
+
+AVG_INFERENCE_TIME = 0.1
+PHYSICAL_CORES_PER_NODE = 4
+
+
+random.seed(170681)
+
+
+class Estimator:
+  def __init__(self, value, variance, smoothing_factor):
+    self.value = value
+    self.variance = variance
+    self.smoothing_factor = smoothing_factor
+
+  def add(self, value, variance, dynamics_changed=False) -> None:
+    process_variance = variance / self.smoothing_factor
+
+    if dynamics_changed or abs(value - self.value)**2 / (self.variance + variance) > 100:
+      process_variance *= self.smoothing_factor
+      print(f'dynamic changed: value={value} state={self.value} +-/ {math.sqrt(self.variance)}')
+
+    gain = (self.variance + process_variance) / (self.variance + process_variance + variance)
+    self.value += gain * (value - self.value)
+    self.variance = (1 - gain) * (self.variance + process_variance)
+
+  def estimate(self) -> float:
+    return self.value
+
+  def lower(self):
+    return max(0.0, self.value - math.sqrt(self.variance))
+
+  def upper(self):
+    return self.value + math.sqrt(self.variance)
+
+
+class Simulator:
+
+  class EventType(enum.IntEnum):
+    ALLOCATION_STARTED = 1
+    REQUEST = 2
+    INFERENCE_COMPLETED = 3
+    MEASURE_RATE = 4
+
+  def __init__(self, get_request_rate):
+    self.get_request_rate = get_request_rate
+
+    # state of the system
+    self.events = []  # contains tuple of (time, type, allocation_id, inference_time)
+    self.num_allocations = 1
+
+    # data for the graphs
+    self.request_rate_data = []
+    self.request_rate_estimates = []
+    self.request_rate_truth = []
+    self.inference_time_data = []
+    self.inference_time_estimates = []
+    self.inference_time_truth = []
+    self.wait_times = []
+    self.queue_sizes = []
+    self.num_allocations_list = []
+    self.num_ml_nodes = []
+    self.est_loads = []
+    self.real_loads = []
+
+  def get_avg_inference_time(self):
+    physical_cores = self.num_allocations // (2 * PHYSICAL_CORES_PER_NODE) * PHYSICAL_CORES_PER_NODE
+    remaining_allocations = self.num_allocations % (2 * PHYSICAL_CORES_PER_NODE)
+    if remaining_allocations < PHYSICAL_CORES_PER_NODE:
+      physical_cores += remaining_allocations
+    else:
+      physical_cores += PHYSICAL_CORES_PER_NODE
+    return AVG_INFERENCE_TIME * self.num_allocations / physical_cores
+
+  def get_random_inference_time(self):
+    return random.uniform(0.5, 1.5) * self.get_avg_inference_time()
+
+  def create_events(self):
+    print('creating events...')
+
+    # initial allocations
+    for alloc_id in range(self.num_allocations):
+      heapq.heappush(self.events, (0, Simulator.EventType.ALLOCATION_STARTED, alloc_id, None))
+
+    # create all requests
+    TIME_STEP = 0.001
+    time = START_TIME + TIME_STEP / 2
+    while time < END_TIME:
+      if random.random() < self.get_request_rate(time) * TIME_STEP:
```

Comment:
This saturates if rate is > 1 / 0.001. I guess it doesn't matter for the parameters you use for simulation, but it would be a little more robust if you generate next time instead as `t += numpy.random.exponential(scale=1/self.get_request_rate(time))`.

Reply:
Reading on I see this will have an impact on implementation of run stat estimation. This could all be resolved, but let's not bother to do that now. I think a comment is warranted that the simulator should only use rate << 1 / TIME_STEP.

Reply:
I've done that before, but it performs badly for low, increasing request rates (the next timestamp is too far in the future, and the rate has increased a lot by then)